corehq/apps/domain/middleware.py

```diff
@@ -49,13 +47,31 @@ class DomainHistoryMiddleware(MiddlewareMixin):
 
     def process_response(self, request, response):
         if hasattr(request, 'domain') and getattr(response, '_remember_domain', True):
-            self.remember_domain_visit(request, response)
+            self.remember_domain_visit(request)
         return response
 
-    def remember_domain_visit(self, request, response):
-        last_visited_domain = request.session.get('last_visited_domain')
-        if last_visited_domain != request.domain:
-            request.session['last_visited_domain'] = request.domain
+    def remember_domain_visit(self, request):
+        if hasattr(request, 'couch_user') and request.couch_user:
+            last_visited_domain = get_last_visited_domain(request.couch_user)
+            if last_visited_domain != request.domain:
+                set_last_visited_domain(request.couch_user, request.domain)
+
+
+def get_last_visited_domain(couch_user):
+    client = cache_core.get_redis_client()
+    return client.get(_last_visited_domain_cache_key(couch_user))
+
+
+def set_last_visited_domain(couch_user, domain):
+    client = cache_core.get_redis_client()
+    cache_expiration = 60 * 60 * 24
+    cache_key = _last_visited_domain_cache_key(couch_user)
+    client.set(cache_key, domain)
+    client.expire(cache_key, cache_expiration)
```

Comment:
I'm not familiar with this pattern, but I do know `client.set` takes a `timeout` argument.  Would that do the same thing in one request?

Reply:
I had just copied this code from elsewhere, so I [looked it up](https://redis.io/commands/expire#how-redis-expires-keys). It seems that the `timeout` param for `set` is slightly different than expiring a key, but for the purposes of this request they're basically the same. The [place I grabbed this code from](https://github.com/dimagi/commcare-hq/blob/0c64140ef61c3e2c5db6ac4ec6e8ac7b8fa7f985/corehq/apps/sms/views.py#L676-L679) is caching a large amount of data, so that's probably why it explicitly expires the cache. Updated in 3f585442bd9